README.md
# Mist Browser

[![Join the chat at https://gitter.im/ethereum/mist](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/ethereum/mist?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![Build Status develop branch](https://travis-ci.org/ethereum/mist.svg?branch=develop)](https://travis-ci.org/ethereum/mist)
[![Code Climate](https://codeclimate.com/github/ethereum/mist/badges/gpa.svg)](https://codeclimate.com/github/ethereum/mist)

The Mist browser is the tool of choice to browse and use Ðapps.

For the Mist API see the [MISTAPI.md](MISTAPI.md).

Please note that this repository is the Electron host for the Meteor based wallet dapp whose repository is located here: https://github.com/ethereum/meteor-dapp-wallet.

## Help and troubleshooting

Please check the [Mist troubleshooting guide](https://github.com/ethereum/mist/wiki).

Or the [Gitter Channel](https://gitter.im/ethereum/mist), to connect with the community for instant help.

## Installation

If you want to install the app from a pre-built version on the [release page](https://github.com/ethereum/mist/releases),
you can simply run the executeable after download.

For updating simply download the new version and copy it over the old one (keep a backup of the old one if you want to be sure).

#### Config folder
The data folder for Mist is stored in other places:

- Windows `%APPDATA%\Mist`
- macOS `~/Library/Application\ Support/Mist`
- Linux `~/.config/Mist`


## Development

For development, a Meteor server will need to be started to assist with live reload and CSS injection.
Once a Mist version is released the Meteor frontend part is bundled using the `meteor-build-client` npm package to create pure static files.

### Dependencies

To run mist in development you need:

- [Node.js](https://nodejs.org) `v7.x` (use the prefered installation method for your OS)
- [Meteor](https://www.meteor.com/install) javascript app framework
- [Yarn](https://yarnpkg.com/) package manager
- [Electron](http://electron.atom.io/) `v1.4.15` cross platform desktop app framework
- [Gulp](http://gulpjs.com/) build and automation system

Install the latter ones via:

    $ curl https://install.meteor.com/ | sh
    $ curl -o- -L https://yarnpkg.com/install.sh | bash
    $ yarn global add electron@1.4.15
    $ yarn global add gulp

### Initialisation

Now you're ready to initialise Mist for development:

    $ git clone https://github.com/ethereum/mist.git
    $ cd mist
    $ yarn

To update Mist in the future, run:

    $ cd mist
    $ git pull
    $ yarn

### Run Mist

For development we start the interface with a Meteor server for autoreload etc.
*Start the interface in a separate terminal window:*

    $ cd mist/interface && meteor --no-release-check

In the original window you can then start Mist with:

    $ cd mist
    $ electron .

*NOTE: client-binaries (e.g. [geth](https://github.com/ethereum/go-ethereum)) specified in [clientBinaries.json](https://github.com/ethereum/mist/blob/master/clientBinaries.json) will be checked during every startup and downloaded if out-of-date, binaries are stored in the [config folder](#config-folder)*

*NOTE: use `--help` to display available options, e.g. `--loglevel debug` (or `trace`) for verbose output*

### Run the Wallet

Start the wallet app for development, *in a separate terminal window:*

    $ cd mist/interface && meteor --no-release-check

    // and in another terminal

    $ cd my/path/meteor-dapp-wallet/app && meteor --port 3050

In the original window you can then start Mist using wallet mode:

    $ cd mist
    $ electron . --mode wallet


### Connecting to node via HTTP instead of IPC

This is useful if you have a node running on another machine, though note that
it's less secure than using the default IPC method.

```bash
$ electron . --rpc http://localhost:8545
```


### Passing options to Geth

You can pass command-line options directly to Geth by prefixing them with `--node-` in
the command-line invocation:

```bash
$ electron . --mode mist --node-rpcport 19343 --node-networkid 2
```

The `--rpc` Mist option is a special case. If you set this to an IPC socket file
path then the `--ipcpath` option automatically gets set, i.e.:

```bash
$ electron . --rpc /my/geth.ipc
```

...is the same as doing...


```bash
$ electron . --rpc /my/geth.ipc --node-ipcpath /my/geth.ipc
```

### Creating a local private net

See this guide to quickly set up a local private network on your computer:
https://gist.github.com/evertonfraga/9d65a9f3ea399ac138b3e40641accf23


### Using Mist with a privatenet

To run a private network you will need to set the IPC path, network id and data
folder:

```bash
$ electron . --rpc ~/Library/Ethereum/geth.ipc --node-networkid 1234 --node-datadir ~/Library/Ethereum/privatenet
```

_NOTE: since `ipcpath` is also a Mist option you do not need to also include a
`--node-ipcpath` option._

You can also launch `geth` separately with the same options prior starting
Mist.


### Deployment

Our build system relies on [gulp](http://gulpjs.com/) and [electron-builder](https://github.com/electron-userland/electron-builder/).

#### Dependencies

[meteor-build-client](https://github.com/frozeman/meteor-build-client) bundles the [meteor](https://www.meteor.com/)-based interface. Install it via:

    $ npm install -g meteor-build-client

Furthermore cross-platform builds require additional [`electron-builder` dependencies](https://github.com/electron-userland/electron-builder/wiki/Multi-Platform-Build#linux). On macOS those are:

    // windows deps
    $ brew install wine --without-x11 mono makensis

    // linux deps
    $ brew install gnu-tar libicns graphicsmagick xz

#### Generate packages

To generate the binaries for Mist run:

    $ gulp

To generate the Ethereum Wallet (this will pack the one Ðapp from https://github.com/ethereum/meteor-dapp-wallet):

    $ gulp --wallet

The generated binaries will be under `dist_mist/release` or `dist_wallet/release`.


#### Options

##### platform

To build binaries for specific platforms (default: all available) use the following flags:

    // on mac
    $ gulp --win --linux --mac

    // on linux
    $ gulp --win --linux

    // on win
    $ gulp --win

##### walletSource

With the `walletSource` you can specify the Wallet branch to use, default is `master`:

    $ gulp --wallet --walletSource develop


Options are:

- `master`
- `develop`
- `local` Will try to build the wallet from [mist/]../meteor-dapp-wallet/app

*Note: applicable only when combined with `--wallet`*

#### Checksums

Spits out the MD5 checksums of distributables.

It expects installer/zip files to be in the generated folders e.g. `dist_mist/release`

    $ gulp checksums [--wallet]


## Testing

Tests are ran using [Spectron](https://github.com/electron/spectron/), a webdriver.io runner built for Electron. 

First make sure to build Mist with:

    $ gulp

Then run the tests:

    $ gulp test

*Note: Integration tests are not yet supported on Windows.*


CHANGELOG.md
# Change Log

## [Unreleased](https://github.com/ethereum/mist/tree/HEAD)

[Full Changelog](https://github.com/ethereum/mist/compare/v0.9.0...HEAD)

**Fixed bugs:**

- Cannot find "delete" icon to delete contract in 0.9.0 [\#2849](https://github.com/ethereum/mist/issues/2849)
- Mist 400.html 404 error page showing blank screen [\#2204](https://github.com/ethereum/mist/issues/2204)

**Closed issues:**

- transactions unconfirmed \ wallet contracts pending creating \ custom tokens not updated?  [\#3012](https://github.com/ethereum/mist/issues/3012)
- I bring to you a mist-ery of invisible funds [\#3010](https://github.com/ethereum/mist/issues/3010)
- ether sending to new wallet problems [\#3008](https://github.com/ethereum/mist/issues/3008)
- sending ether to new ether wallet problems [\#3007](https://github.com/ethereum/mist/issues/3007)
- ethereum wallet not sending correctly [\#3006](https://github.com/ethereum/mist/issues/3006)
- deleted single chaindata file and now cant connect [\#2991](https://github.com/ethereum/mist/issues/2991)
- permanent solution ? [\#2971](https://github.com/ethereum/mist/issues/2971)
- Ethereum Wallet - Tokens are missing... 0.9.0 [\#2968](https://github.com/ethereum/mist/issues/2968)
- e [\#2965](https://github.com/ethereum/mist/issues/2965)
- ETH don't sync all blocks [\#2957](https://github.com/ethereum/mist/issues/2957)
- Ethereum Wallet never stops syncing \(3 days still not synced!\) [\#2956](https://github.com/ethereum/mist/issues/2956)
- geth 1.6.7 OOM after 2 mins in windows-386/go1.8.3 [\#2954](https://github.com/ethereum/mist/issues/2954)
- Can I Erase?  [\#2943](https://github.com/ethereum/mist/issues/2943)
- Mist Version 0.9.0 , Win 7, eth/geth  - I installed mist 2 month ago, can´t find my password [\#2941](https://github.com/ethereum/mist/issues/2941)
- Impossible to install on Windows 10 [\#2940](https://github.com/ethereum/mist/issues/2940)
- Lost Ether from CEX.IO [\#2932](https://github.com/ethereum/mist/issues/2932)
- LTC -\> Ethereum via LRPHpZ1zN9yH6afn2oKsawhGLZNKXRv26m ? [\#2929](https://github.com/ethereum/mist/issues/2929)
- ICN transfer missing [\#2928](https://github.com/ethereum/mist/issues/2928)
- When I try to log into my Parity, it gives me the option to only create a new account, please help [\#2927](https://github.com/ethereum/mist/issues/2927)
- New geth installation hangs up on block 2675000 and out of gas warning [\#2925](https://github.com/ethereum/mist/issues/2925)
- Key [\#2922](https://github.com/ethereum/mist/issues/2922)
- Wallet is getting too big \(80+ GB and growing fast\). Any way to reduce the size? [\#2915](https://github.com/ethereum/mist/issues/2915)
- Wrong password but I'm sure it's ok! [\#2910](https://github.com/ethereum/mist/issues/2910)
- EMAIL SCAM ALERT - MIST SUPPORT [\#2908](https://github.com/ethereum/mist/issues/2908)
- Explicit "storage" keyword [\#2903](https://github.com/ethereum/mist/issues/2903)
- Ethereum-Wallet-macosx-0-9-0.dmg checksum correct?? [\#2889](https://github.com/ethereum/mist/issues/2889)
- Mist and Ethereum-Wallet checksums for v0.9.0 are MD5 but should be SHA256 [\#2887](https://github.com/ethereum/mist/issues/2887)
- contracts and history gone, after downloading new version [\#2885](https://github.com/ethereum/mist/issues/2885)
- Double Spend Affecting the Wallet? [\#2883](https://github.com/ethereum/mist/issues/2883)
- Contract not working [\#2881](https://github.com/ethereum/mist/issues/2881)
- Can't start parity [\#2880](https://github.com/ethereum/mist/issues/2880)
- Cant start parity  [\#2879](https://github.com/ethereum/mist/issues/2879)
- \#\#\#\#\#\#\#\#\#\# BAD BLOCK \#\#\#\#\#\#\#\#\# Chain config: {ChainID: 1 Homestead: 1150000 DAO: 1920000 DAOSupport: true EIP150: 2463000 EIP155: 2675000 EIP158: 2675000 Metropolis: 9223372036854775807 Engine: ethash}  Number: 4071855 Hash: 0xe757a526b8ce62ebff1e43dc8bb2ddb5a1cc5f906e6fedc34bef26402103bf01 [\#2876](https://github.com/ethereum/mist/issues/2876)
- QR-code generator swaps cases of some symbols  [\#2875](https://github.com/ethereum/mist/issues/2875)
- Please add SHA-256 hashes for 0.9.0 [\#2873](https://github.com/ethereum/mist/issues/2873)
- How to send 0.0 ETH  [\#2871](https://github.com/ethereum/mist/issues/2871)
- Can i delete the blockchain lokal without loose my coins? [\#2870](https://github.com/ethereum/mist/issues/2870)
- Popup windows issue [\#2869](https://github.com/ethereum/mist/issues/2869)
- Sync hangs at 99% 315 Blocks left.. [\#2865](https://github.com/ethereum/mist/issues/2865)
- PHISING SCAM STORY $ 100 K + [\#2858](https://github.com/ethereum/mist/issues/2858)
- Failed to close database                 database=C:\\Users\\Link\\AppData\\Roaming\\Ethereum\\geth\\chaindata err="leveldb/table: corruption on data-block \(pos=195003226\): checksum mismatch, want=0x8d32caa6 got=0x5559fd50 \[file=2089555.ldb\]" [\#2857](https://github.com/ethereum/mist/issues/2857)
- PHISHING SCAM  [\#2855](https://github.com/ethereum/mist/issues/2855)
- Couldn't connect to node? [\#2854](https://github.com/ethereum/mist/issues/2854)
- ALL my Coins have been taken away overnight [\#2853](https://github.com/ethereum/mist/issues/2853)
- Navigating bzz from mist? [\#2852](https://github.com/ethereum/mist/issues/2852)
- Memory usage geth.exe [\#2846](https://github.com/ethereum/mist/issues/2846)
- Still missing eth in wallet from 45 days ago  [\#2844](https://github.com/ethereum/mist/issues/2844)
- Bundled swarm uses lots of bandwidth [\#2843](https://github.com/ethereum/mist/issues/2843)
- Mist 0.9.0 missing eth in web3 [\#2841](https://github.com/ethereum/mist/issues/2841)
- No Blocks Synced after update to 0.9.0 [\#2837](https://github.com/ethereum/mist/issues/2837)
- ETH still not on account ballance  [\#2836](https://github.com/ethereum/mist/issues/2836)
- Problem in withdraw from wallet contract [\#2835](https://github.com/ethereum/mist/issues/2835)

main.js
global._ = require('./modules/utils/underscore');
const { app, dialog, ipcMain, shell, protocol } = require('electron');
const timesync = require('os-timesync');
const dbSync = require('./modules/dbSync.js');
const i18n = require('./modules/i18n.js');
const logger = require('./modules/utils/logger');
const Sockets = require('./modules/socketManager');
const Windows = require('./modules/windows');
const ClientBinaryManager = require('./modules/clientBinaryManager');
const UpdateChecker = require('./modules/updateChecker');
const Settings = require('./modules/settings');
const Q = require('bluebird');
const windowStateKeeper = require('electron-window-state');

Q.config({
    cancellation: true,
});


// logging setup
const log = logger.create('main');


Settings.init();

if (Settings.cli.version) {
    log.info(Settings.appVersion);

    process.exit(0);
}

if (Settings.cli.ignoreGpuBlacklist) {
    app.commandLine.appendSwitch('ignore-gpu-blacklist', 'true');
}

if (Settings.inAutoTestMode) {
    log.info('AUTOMATED TESTING');
}

log.info(`Running in production mode: ${Settings.inProductionMode}`);

if (Settings.rpcMode === 'http') {
    log.warn('Connecting to a node via HTTP instead of ipcMain. This is less secure!!!!'.toUpperCase());
}

// db
const db = global.db = require('./modules/db');


require('./modules/ipcCommunicator.js');
const appMenu = require('./modules/menuItems');
const ipcProviderBackend = require('./modules/ipc/ipcProviderBackend.js');
const ethereumNode = require('./modules/ethereumNode.js');
const swarmNode = require('./modules/swarmNode.js');
const nodeSync = require('./modules/nodeSync.js');

global.webviews = [];

global.mining = false;

global.icon = `${__dirname}/icons/${Settings.uiMode}/icon.png`;
global.mode = Settings.uiMode;
global.dirname = __dirname;

global.i18n = i18n;


// INTERFACE PATHS
// - WALLET
if (Settings.uiMode === 'wallet') {
    log.info('Starting in Wallet mode');

    global.interfaceAppUrl = (Settings.inProductionMode)
        ? `file://${__dirname}/interface/wallet/index.html`
        : 'http://localhost:3050';
    global.interfacePopupsUrl = (Settings.inProductionMode)
        ? `file://${__dirname}/interface/index.html`
        : 'http://localhost:3000';

// - MIST
} else {
    log.info('Starting in Mist mode');

    let url = (Settings.inProductionMode)
        ? `file://${__dirname}/interface/index.html`
        : 'http://localhost:3000';

    if (Settings.cli.resetTabs) {
        url += '?reset-tabs=true';
    }

    global.interfaceAppUrl = global.interfacePopupsUrl = url;
}


// prevent crashed and close gracefully
process.on('uncaughtException', (error) => {
    log.error('UNCAUGHT EXCEPTION', error);
    app.quit();
});


// Quit when all windows are closed.
app.on('window-all-closed', () => {
    app.quit();
});

// Listen to custom protocole incoming messages, needs registering of URL schemes
app.on('open-url', (e, url) => {
    log.info('Open URL', url);
});


let killedSocketsAndNodes = false;

app.on('before-quit', (event) => {
    if (!killedSocketsAndNodes) {
        log.info('Defer quitting until sockets and node are shut down');

        event.preventDefault();

        // sockets manager
        Sockets.destroyAll()
            .catch(() => {
                log.error('Error shutting down sockets');
            });

        // delay quit, so the sockets can close
        setTimeout(() => {
            ethereumNode.stop()
            .then(() => {
                killedSocketsAndNodes = true;

                return db.close();
            })
            .then(() => {
                app.quit();
            });
        }, 500);
    } else {
        log.info('About to quit...');
    }
});


let mainWindow;
let splashWindow;
let onReady;
let startMainWindow;

// This method will be called when Electron has done everything
// initialization and ready for creating browser windows.
app.on('ready', () => {
    // if using HTTP RPC then inform user
    if (Settings.rpcMode === 'http') {
        dialog.showErrorBox('Insecure RPC connection', `
WARNING: You are connecting to an Ethereum node via: ${Settings.rpcHttpPath}

This is less secure than using local IPC - your passwords will be sent over the wire in plaintext.

Only do this if you have secured your HTTP connection or you know what you are doing.
`);
    }

    // initialise the db
    global.db.init().then(onReady).catch((err) => {
        log.error(err);
        app.quit();
    });
});

// Allows the Swarm protocol to behave like http
protocol.registerStandardSchemes(['bzz']);

onReady = () => {
    global.config = db.getCollection('SYS_config');

    // setup DB sync to backend
    dbSync.backendSyncInit();

    // Initialise window mgr
    Windows.init();

    // Enable the Swarm protocol
    protocol.registerHttpProtocol('bzz', (request, callback) => {
        const redirectPath = `${Settings.swarmURL}/${request.url.replace('bzz:/', 'bzz://')}`;
        callback({ method: request.method, referrer: request.referrer, url: redirectPath });
    }, (error) => {
        if (error) {
            log.error(error);
        }
    });

    // check for update
    if (!Settings.inAutoTestMode) UpdateChecker.run();

    // initialize the web3 IPC provider backend
    ipcProviderBackend.init();

    // instantiate custom protocols
    // require('./customProtocols.js');

    // change to user language now that global.config object is ready
    i18n.changeLanguage(Settings.language);

    // add menu already here, so we have copy and past functionality
    appMenu();

    // Create the browser window.

    const defaultWindow = windowStateKeeper({
        defaultWidth: 1024 + 208,
        defaultHeight: 720
    });

    // MIST
    if (Settings.uiMode === 'mist') {
        mainWindow = Windows.create('main', {
            primary: true,
            electronOptions: {
                width: Math.max(defaultWindow.width, 500),
                height: Math.max(defaultWindow.height, 440),
                x: defaultWindow.x,
                y: defaultWindow.y,
                webPreferences: {
                    nodeIntegration: true, /* necessary for webviews;
                        require will be removed through preloader */
                    preload: `${__dirname}/modules/preloader/mistUI.js`,
                    'overlay-fullscreen-video': true,
                    'overlay-scrollbars': true,
                    experimentalFeatures: true,
                },
            },
        });

    // WALLET
    } else {
        mainWindow = Windows.create('main', {
            primary: true,
            electronOptions: {
                width: Math.max(defaultWindow.width, 500),
                height: Math.max(defaultWindow.height, 440),
                x: defaultWindow.x,
                y: defaultWindow.y,
                webPreferences: {
                    preload: `${__dirname}/modules/preloader/walletMain.js`,
                    'overlay-fullscreen-video': true,
                    'overlay-scrollbars': true,
                },
            },
        });
    }

    // Delegating events to save window bounds on windowStateKeeper
    defaultWindow.manage(mainWindow.window);

    if (!Settings.inAutoTestMode) {
        splashWindow = Windows.create('splash', {
            primary: true,
            url: `${global.interfacePopupsUrl}#splashScreen_${Settings.uiMode}`,
            show: true,
            electronOptions: {
                width: 400,
                height: 230,
                resizable: false,
                backgroundColor: '#F6F6F6',
                useContentSize: true,
                frame: false,
                webPreferences: {
                    preload: `${__dirname}/modules/preloader/splashScreen.js`,
                },
            },
        });
    }

    // check time sync
    // var ntpClient = require('ntp-client');
    // ntpClient.getNetworkTime("pool.ntp.org", 123, function(err, date) {
    timesync.checkEnabled((err, enabled) => {
        if (err) {
            log.error('Couldn\'t get time from NTP time sync server.', err);
            return;
        }

        if (!enabled) {
            dialog.showMessageBox({
                type: 'warning',
                buttons: ['OK'],
                message: global.i18n.t('mist.errors.timeSync.title'),
                detail: `${global.i18n.t('mist.errors.timeSync.description')}\n\n${global.i18n.t(`mist.errors.timeSync.${process.platform}`)}`,
            }, () => {
            });
        }
    });


    const kickStart = () => {
        // client binary stuff
        ClientBinaryManager.on('status', (status, data) => {
            Windows.broadcast('uiAction_clientBinaryStatus', status, data);
        });

        // node connection stuff
        ethereumNode.on('nodeConnectionTimeout', () => {
            Windows.broadcast('uiAction_nodeStatus', 'connectionTimeout');
        });

        ethereumNode.on('nodeLog', (data) => {
            Windows.broadcast('uiAction_nodeLogText', data.replace(/^.*[0-9]]/, ''));
        });

        // state change
        ethereumNode.on('state', (state, stateAsText) => {
            Windows.broadcast('uiAction_nodeStatus', stateAsText,
                ethereumNode.STATES.ERROR === state ? ethereumNode.lastError : null
            );
        });

        // starting swarm
        swarmNode.on('starting', () => {
            Windows.broadcast('uiAction_swarmStatus', 'starting');
        });

        // swarm download progress
        swarmNode.on('downloadProgress', (progress) => {
            Windows.broadcast('uiAction_swarmStatus', 'downloadProgress', progress);
        });

        // started swarm
        swarmNode.on('started', (isLocal) => {
            Windows.broadcast('uiAction_swarmStatus', 'started', isLocal);
        });


        // capture sync results
        const syncResultPromise = new Q((resolve, reject) => {
            nodeSync.on('nodeSyncing', (result) => {
                Windows.broadcast('uiAction_nodeSyncStatus', 'inProgress', result);
            });

            nodeSync.on('stopped', () => {
                Windows.broadcast('uiAction_nodeSyncStatus', 'stopped');
            });

            nodeSync.on('error', (err) => {
                log.error('Error syncing node', err);

                reject(err);
            });

            nodeSync.on('finished', () => {
                nodeSync.removeAllListeners('error');
                nodeSync.removeAllListeners('finished');

                resolve();
            });
        });

        // check legacy chain
        // CHECK for legacy chain (FORK RELATED)
        Q.try(() => {
            // open the legacy chain message
            if ((Settings.loadUserData('daoFork') || '').trim() === 'false') {
                dialog.showMessageBox({
                    type: 'warning',
                    buttons: ['OK'],
                    message: global.i18n.t('mist.errors.legacyChain.title'),
                    detail: global.i18n.t('mist.errors.legacyChain.description')
                }, () => {
                    shell.openExternal('https://github.com/ethereum/mist/releases');
                    app.quit();
                });

                throw new Error('Cant start client due to legacy non-Fork setting.');
            }
        })
        .then(() => {
            return ClientBinaryManager.init();
        })
        .then(() => {
            return ethereumNode.init();
        })
        .then(() => {
            // Wallet shouldn't start Swarm
            if (Settings.uiMode === 'wallet') {
                return Promise.resolve();
            }
            return swarmNode.init();
        })
        .then(function sanityCheck() {
            if (!ethereumNode.isIpcConnected) {
                throw new Error('Either the node didn\'t start or IPC socket failed to connect.');
            }

            /* At this point Geth is running and the socket is connected. */
            log.info('Connected via IPC to node.');

            // update menu, to show node switching possibilities
            appMenu();
        })
        .then(function getAccounts() {
            return ethereumNode.send('eth_accounts', []);
        })
        .then(function onboarding(resultData) {

            if (ethereumNode.isGeth && (resultData.result === null || (_.isArray(resultData.result) && resultData.result.length === 0))) {
                log.info('No accounts setup yet, lets do onboarding first.');

                return new Q((resolve, reject) => {
                    const onboardingWindow = Windows.createPopup('onboardingScreen', {
                        primary: true,
                        electronOptions: {
                            width: 576,
                            height: 442,
                        },
                    });

                    onboardingWindow.on('closed', () => {
                        app.quit();
                    });

                    // change network types (mainnet, testnet)
                    ipcMain.on('onBoarding_changeNet', (e, testnet) => {
                        const newType = ethereumNode.type;
                        const newNetwork = testnet ? 'rinkeby' : 'main';

                        log.debug('Onboarding change network', newType, newNetwork);

                        ethereumNode.restart(newType, newNetwork)
                            .then(function nodeRestarted() {
                                appMenu();
                            })
                            .catch((err) => {
                                log.error('Error restarting node', err);

                                reject(err);
                            });
                    });

                    // launch app
                    ipcMain.on('onBoarding_launchApp', () => {
                        // prevent that it closes the app
                        onboardingWindow.removeAllListeners('closed');
                        onboardingWindow.close();

                        ipcMain.removeAllListeners('onBoarding_changeNet');
                        ipcMain.removeAllListeners('onBoarding_launchApp');

                        resolve();
                    });

                    if (splashWindow) {
                        splashWindow.hide();
                    }
                });
            }

            return;
        })
        .then(function doSync() {
            // we're going to do the sync - so show splash
            if (splashWindow) {
                splashWindow.show();
            }

            if (!Settings.inAutoTestMode) {
                return syncResultPromise;
            }

            return;
        })
        .then(function allDone() {
            startMainWindow();
        })
        .catch((err) => {
            log.error('Error starting up node and/or syncing', err);
        }); /* socket connected to geth */
    }; /* kick start */

    if (splashWindow) {
        splashWindow.on('ready', kickStart);
    } else {
        kickStart();
    }
}; /* onReady() */


/**
Start the main window and all its processes

@method startMainWindow
*/
startMainWindow = () => {
    log.info(`Loading Interface at ${global.interfaceAppUrl}`);

    mainWindow.on('ready', () => {
        if (splashWindow) {
            splashWindow.close();
        }

        mainWindow.show();
    });

    mainWindow.load(global.interfaceAppUrl);

    // close app, when the main window is closed
    mainWindow.on('closed', () => {
        app.quit();
    });

    // observe Tabs for changes and refresh menu
    const Tabs = global.db.getCollection('UI_tabs');

    const sortedTabs = Tabs.addDynamicView('sorted_tabs');
    sortedTabs.applySimpleSort('position', false);

    const refreshMenu = () => {
        clearTimeout(global._refreshMenuFromTabsTimer);

        global._refreshMenuFromTabsTimer = setTimeout(() => {
            log.debug('Refresh menu with tabs');

            global.webviews = sortedTabs.data();

            appMenu(global.webviews);
        }, 1000);
    };

    Tabs.on('insert', refreshMenu);
    Tabs.on('update', refreshMenu);
    Tabs.on('delete', refreshMenu);
};
